Improve regex optimizations around repetitions of the same character

stephentoub opened the pull request:
- For concatenations, we currently do two passes over its nodes: one that coalesces adjacent characters into strings, and one that coalesces adjacent loops.  This swaps the order so that we first coalesce adjacent loops.  We want to coalesce loops first so that we prioritize associating a One with an adjacent loop rather than with an adjacent string.  This is primarily beneficial later on for auto-atomicity.  Consider `a+ab` composed of a loop, a One 'a', and a One 'b'.  We're either going to make that into a loop `a{2,}` followed by `b` or into a loop `a+` followed by the multi `ab`.  We want to do the former, as then when we apply auto-atomicity to the `a{2,}`, we'll see that it's followed by something non-overlapping ('b'), and make the loop atomic.  In constrast, if we joined the `a` and `b`, then auto-atomicity for the `a+` would see it's followed by an `a` and it wouldn't be upgraded.
- Currently, the adjacent loop coalescing logic joins two individual One nodes with the same Ch value into a repeater.  That's actually a deoptimization, so this stops doing so.  We'd rather have `aa` be evaluated as a Multi of two characters rather than as an `a{2}` repeater, as we're able to apply better optimizations with multis, e.g. taking advantage of StartsWith for matching.
- Also in the adjacent loop logic, we're already checking for a loop followed by a one but we're not checking for a loop followed by a multi.  For the same auto-atomicity benefits discussed earlier, we want to shift any adjacent text from a multi following a loop back into the loop, as it makes it more likely we'll then be able to upgrade that loop to atomic. (We could also add logic for a multi followed by a loop, but the benefits there are less obvious.)
- When creating repeaters, we're actually better off creating multis, for the same reasons outlined earlier.  But we don't want to create massive strings in the super rare case where large repeaters for single chars are used, so we only do so up to a limit.
- When emitting generated code for a One repeater, which might be used to implement the required portion of other loops, emit it as a multi match for short enough repetitions.

Fixes https://github.com/dotnet/runtime/issues/64991

am11 commented on src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:
```diff
@@ -476,7 +486,6 @@ public void PatternsReduceIdentically(string pattern1, string pattern2)
 
         [Theory]
         // Not coalescing loops
-        [InlineData("aa", "a{2}")]
```
(Unrelated to this change) In these test methods, is there a reason why we chose not to use `Assert.Equal/NotEqual(result1, result2)` instead of manual `throw`?

stephentoub replied:
I'm sure there was at the time.  At present I can't explain it :)

stephentoub replied:
Oh, I remember why.  The built-in Assert.{Not}Equals trims the inputs to just show the first location that didn't match, plus a few characters around that.  That's actively harmful for debugging failures here.

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -2412,11 +2478,31 @@ public bool TryGetJoinableLengthCheckChildRange(int childIndex, out int required
+                    case <= MultiVsRepeaterLimit when Kind == RegexNodeKind.One:
+                        // The same character is repeated a fixed number of times, so it's actually a multi.
+                        // While this could remain a repeater, multis are more readily optimized later in
+                        // processing,The counts used here in real-world expressions are invariably small (e.g. 4),
```
tpoy

danmoseley commented on src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:
```diff
@@ -124,6 +124,16 @@ private static int GetMinRequiredLength(Regex r)
         [InlineData("aa+", "a{2,}")]
         [InlineData("aa?", "a{1,2}")]
         [InlineData("aa{1,3}", "a{2,4}")]
+        [InlineData("aaa+b", "aa(?>a+)b")]
```
would this also reduce the same as `(?>aaa)a*b`

stephentoub replied:
No:
https://github.com/dotnet/runtime/pull/65273/files#diff-69d0a75aaf0674d30e057b6ac1d37203ee4757db089bac19f32a1d31d52a10c9R1689-R1692

danmoseley commented on src/libraries/System.Text.RegularExpressions/tests/RegexReductionTests.cs:
```diff
@@ -124,6 +124,16 @@ private static int GetMinRequiredLength(Regex r)
         [InlineData("aa{1,3}", "a{2,4}")]
+        [InlineData("aaa+b", "aa(?>a+)b")]
+        [InlineData("a+aab", "(?>a{3,})b")]
+        [InlineData("a{3}b", "aaab")]
```
This test only verifies that they both reduce the same. Your change avoids reducing `aaa` into `a{3}` but this is only testing that they're equivalent. Is that useful ? -- I imagine 1000 tests would fail if it wasn't true.

stephentoub replied:
> Your change avoids reducing aaa into a{3} but this is only testing that they're equivalent

This is testing that they reduce to the same thing, i.e. that the node trees they produce are identical.  That's not the same as their having the same execution semantics.

danmoseley replied:
Right, but it would pass whether the node tree produced had a loop or had a multi. And this change is designed to ensure it ends up as the right one. Both trees will also produce the same results for the user. The test is obviously worth having -- just wondering aloud about whether we need a new kind of test that verifies some aspect of the node tree.

stephentoub replied:
> Right, but it would pass whether the node tree produced had a loop or had a multi

Yes, it's not verifying the direction it takes, just that they reduce to the same thing.  If that's what you were saying, I'm not sure what you meant about "I imagine 1000 tests would fail if it wasn't true".

danmoseley replied:
Sorry, my communication is poor. I figure there's 3 aspects

* does expression X1 reduce to the same node tree as expression X2 -- these tests.
* is that node tree structured as ABC, or have property D 
* does expression X functionally behave as expected --- this is what we have 1000's of tests for.

I'm wondering whether, if I reversed this PR so that multi was collapsed into a loop or whatnot, we'd see test failures, because we do'nt have the 2nd category

danmoseley replied:
Eg., if the node tree was serialized into a descriptive string (debug only) then a test could verify that expression X formed serialized node tree S.

stephentoub replied:
These expressions behave identically and reduce to the same thing; the only thing that would be different based on which direction they reduced would be their execution performance.  We do not currently have tests that validate the tree it exactly the set of nodes we expect it to be, just these that validate two expressions parse into the same exact tree.

danmoseley replied:
right, agreed.

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -14,6 +14,8 @@ internal sealed class RegexNode
         /// <summary>empty bit from the node's options to store data on whether a node contains captures</summary>
         internal const RegexOptions HasCapturesFlag = (RegexOptions)(1 << 31);
+        /// <summary>Arbitrary number of repetitions of the same character when we'd prefer to represent that as a repeater of that character rather than a string.</summary>
+        internal const int MultiVsRepeaterLimit = 64;
```
How did you come up with this - entirely arbitrarily? 😛

stephentoub replied:
> entirely arbitrarily?

Somewhat arbitrary.  In our corpus of regexes, the largest single-char repeater has a length of 23.  I chose something significantly larger than that while still being a reasonable length for a string to create.

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1626,6 +1642,55 @@ static bool CanCombineCounts(int nodeMin, int nodeMax, int nextMin, int nextMax)
+                            }
+                            break;
+
+                        // NOTE: We could add support for coalescing a string with a subsequent loop, but the benefits of that
+                        // are limited. Pulling a subsequent string's prefix back into the loop helps with making the loop atomic,
+                        // but if the loop is after the string, pulling the suffix of the string forward into the loop may actually
+                        // be a deoptimization as those characters could end up matching more slowly as part of loop matching.
```
could you clarify "those characters could end up matching more slowly as part of loop matching." ?

stephentoub replied:
A loop gets emitted as iterating one char at a time, along with extra tracking boilerplate, whereas a string gets emitted as an unrolled/vectorized comparison of the string.  Consider the pattern `aaaaa{10,}\w`.  That gets emitted today as:
```C#
                    // Match the string "aaaa".
                    {
                        byteSpan = global::System.Runtime.InteropServices.MemoryMarshal.AsBytes(slice);
                        if ((uint)slice.Length < 4 ||
                            global::System.Buffers.Binary.BinaryPrimitives.ReadUInt64LittleEndian(byteSpan) != 0x61006100610061ul)
                        {
                            goto NoMatch;
                        }
                    }
                    
                    // Match 'a' greedily at least 10 times.
                    //{
                        pos += 4;
                        slice = slice.Slice(4);
                        charloop_starting_pos = pos;
                        
                        int iteration = 0;
                        while ((uint)iteration < (uint)slice.Length && slice[iteration] == 'a')
                        {
                            iteration++;
                        }
                        
                        if (iteration < 10)
                        {
                            goto NoMatch;
                        }
                        
                        slice = slice.Slice(iteration);
                        pos += iteration;
                        
                        charloop_ending_pos = pos;
                        charloop_starting_pos += 10;
                        goto CharLoopEnd;
                        
                        CharLoopBacktrack:
                        StackPop2(base.runstack, ref stackpos, out charloop_ending_pos, out charloop_starting_pos);
                        
                        if (charloop_starting_pos >= charloop_ending_pos)
                        {
                            goto NoMatch;
                        }
                        pos = --charloop_ending_pos;
                        slice = inputSpan.Slice(pos, end - pos);
                        
                        CharLoopEnd:
                        StackPush2(ref base.runstack!, ref stackpos, charloop_starting_pos, charloop_ending_pos);
                    //}
                    
                    // Match a word character.
                    {
                        if (slice.IsEmpty || !IsWordChar(slice[0]))
                        {
                            goto CharLoopBacktrack;
                        }
                    }
```
The comparison of the first four `a`s happens in a single read/comparison.  If the string were instead coalesced into the loop and it were emitted as `a{14,}\w`, each of those four characters would then be iterated once by one rather than compared in bulk.  So we'd essentially be paying more per character to compare them as part of the repeater rather than as part of the string.  It doesn't seem to be a very measurable difference, which is also why I didn't bother to do the opposite either and try to separate out the min iterations into a multi that comes before it.  These situations also aren't very common.  I think it's worth it for the case I added of a loop followed by a multi, as the impact of removing the backtracking can be huge, but they don't show up very often (there's only 26 patterns in the whole corpus that contain a Oneloop with a lower bound > 1 and an upper bound greater than the lower bound), so the difference in the comparison cost isn't all that much of a motivation.

joperezr commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1453,13 +1467,15 @@ private RegexNode ReduceConcatenation()
 
+            // Coalesce adjacent characters/strings.  This is done after the adjacent loop coalescing so that
+            // a One adjacent to both a Multi and a Loop prefers being folded into the Loop rather than into
+            // the Multi.  Doing so helps with auto-atomicity when it's later applied.
```
NIT: Don't reset CI for this, but it would be nice if we also add to these comments a concrete example, like: `a+ab` => `(?>a{2,})b`

stephentoub replied:
There are examples in the XML comments on the methods being called.

joperezr commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -1626,6 +1642,55 @@ static bool CanCombineCounts(int nodeMin, int nodeMax, int nextMin, int nextMax)
                             break;
 
+                        // Coalescing a loop with a subsequent string
+                        case RegexNodeKind.Oneloop or RegexNodeKind.Oneloopatomic or RegexNodeKind.Onelazy when nextNode.Kind == RegexNodeKind.Multi && currentNode.Ch == nextNode.Str![0]:
```
Should this optimization also happen when nextNode.Kind == One and the character matches?

stephentoub replied:
Do you mean like https://github.com/dotnet/runtime/pull/65273/files/209eef017459fe2f2c6a697b62d4d2d31ef8c76a#diff-69d0a75aaf0674d30e057b6ac1d37203ee4757db089bac19f32a1d31d52a10c9R1629-R1630 ?

joperezr replied:
oh, I missed that, yeah that is what I meant.

stephentoub committed:
Improve regex optimizations around repetitions of the same character

- For concatenations, we currently do two passes over its nodes: one that coalesces adjacent characters into strings, and one that coalesces adjacent loops.  This swaps the order so that we first coalesce adjacent loops.  We want to coalesce loops first so that we prioritize associating a One with an adjacent loop rather than with an adjacent string.  This is primarily beneficial later on for auto-atomicity.  Consider `a+ab` composed of a loop, a One 'a', and a One 'b'.  We're either going to make that into a loop `a{2,}` followed by `b` or into a loop `a+` followed by the multi `ab`.  We want to do the former, as then when we apply auto-atomicity to the `a{2,}`, we'll see that it's followed by something non-overlapping ('b'), and make the loop atomic.  In constrast, if we joined the `a` and `b`, then auto-atomicity for the `a+` would see it's followed by an `a` and it wouldn't be upgraded.
- Currently, the adjacent loop coalescing logic joins two individual One nodes with the same Ch value into a repeater.  That's actually a deoptimization, so this stops doing so.  We'd rather have `aa` be evaluated as a Multi of two characters rather than as an `a{2}` repeater, as we're able to apply better optimizations with multis, e.g. taking advantage of StartsWith for matching.
- Also in the adjacent loop logic, we're already checking for a loop followed by a one but we're not checking for a loop followed by a multi.  For the same auto-atomicity benefits discussed earlier, we want to shift any adjacent text from a multi following a loop back into the loop, as it makes it more likely we'll then be able to upgrade that loop to atomic. (We could also add logic for a multi followed by a loop, but the benefits there are less obvious.)
- When creating repeaters, we're actually better off creating multis, for the same reasons outlined earlier.  But we don't want to create massive strings in the super rare case where large repeaters for single chars are used, so we only do so up to a limit.
- When emitting generated code for a One repeater, which might be used to implement the required portion of other loops, emit it as a multi match for short enough repetitions.
